library/core/src/ffi/va_list.rs

```diff
@@ -158,52 +120,41 @@ pub struct VaList<'a, 'f: 'a> {
         not(target_family = "wasm"),
         not(target_os = "uefi"),
         not(windows),
-    ))]
-    inner: &'a mut VaListImpl<'f>,
+    ) => {
+        /// A wrapper for a `va_list`
+        #[repr(transparent)]
+        #[derive(Debug)]
+        pub struct VaList<'a, 'f: 'a> {
+            inner: &'a mut VaListImpl<'f>,
+            _marker: PhantomData<&'a mut VaListImpl<'f>>,
+        }
 
-    _marker: PhantomData<&'a mut VaListImpl<'f>>,
-}
 
-#[cfg(any(
-    all(
-        not(target_arch = "aarch64"),
-        not(target_arch = "powerpc"),
-        not(target_arch = "s390x"),
-        not(target_arch = "x86_64")
-    ),
-    target_arch = "xtensa",
-    all(target_arch = "aarch64", target_vendor = "apple"),
-    target_family = "wasm",
-    target_os = "uefi",
-    windows,
-))]
-impl<'f> VaListImpl<'f> {
-    /// Converts a `VaListImpl` into a `VaList` that is binary-compatible with C's `va_list`.
-    #[inline]
-    pub fn as_va_list<'a>(&'a mut self) -> VaList<'a, 'f> {
-        VaList { inner: VaListImpl { ..*self }, _marker: PhantomData }
+        impl<'f> VaListImpl<'f> {
+            /// Converts a [`VaListImpl`] into a [`VaList`] that is binary-compatible with C's `va_list`.
+            #[inline]
+            pub fn as_va_list<'a>(&'a mut self) -> VaList<'a, 'f> {
+                VaList { inner: self, _marker: PhantomData }
+            }
+        }
     }
-}
 
-#[cfg(all(
-    any(
-        target_arch = "aarch64",
-        target_arch = "powerpc",
-        target_arch = "s390x",
-        target_arch = "xtensa",
-        target_arch = "x86_64"
-    ),
-    not(target_arch = "xtensa"),
-    any(not(target_arch = "aarch64"), not(target_vendor = "apple")),
-    not(target_family = "wasm"),
-    not(target_os = "uefi"),
-    not(windows),
-))]
-impl<'f> VaListImpl<'f> {
-    /// Converts a `VaListImpl` into a `VaList` that is binary-compatible with C's `va_list`.
-    #[inline]
-    pub fn as_va_list<'a>(&'a mut self) -> VaList<'a, 'f> {
-        VaList { inner: self, _marker: PhantomData }
+    _ => {
+        /// A wrapper for a `va_list`
+        #[repr(transparent)]
+        #[derive(Debug)]
+        pub struct VaList<'a, 'f: 'a> {
+            inner: VaListImpl<'f>,
```

Comment:
I'm confused by this implementation that is used when we don't know the layout of `va_list` (the `_ => { ... }` case in the `cfg_match!` above). It is inconsistent with the others. 

Could we not have `&'a mut VaListImpl<'f>` here like above, and make the `VaListImpl` a ZST containing just the invariance marker?

Is that fallback implementation even meaningful?

Reply:
Ok no, after reading a bit more there are just two approaches 

1. `va_list` is just an opaque pointer (probably just the next variadic argument on the stack, though it's opaque so we can't know)
2. `va_list` is some sort of struct. It looks like this is used when registers are also at play for passing variadic arguments

These two approaches are captured by the two different ways of representing `VaList`. I added a comment with some of that information, in particular https://github.com/rust-lang/rust/pull/56599 had some useful info.

Reply:
That does raise the question: what is `VaList` for? it has no methods of its own, so I don't see its purpose.

Reply:
`VaList` is supposed to be ABI compatible with C's `va_list`

Reply:
Yes, I was confused by the C definition, until I rediscovered that nasty thing where `sizeof` in C is not the same as `size_of` in rust: for arrays, `sizeof` gives the length of the array, though it is really used/passed as a pointer type.

Anyway, https://github.com/rust-lang/rust/issues/44930#issuecomment-2902519131 has an updated design for this api, which keeps `VaListImpl` private.